Add unsigned support to `IValue`

malfet opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #160102

- Moved repeated logic of saving int64/uint64 into a polymorphic container into `THPUtils_unpackInteger`
- Added `TestPythonDispatch.test_dispatch_uint64` regression test 

Fixes https://github.com/pytorch/pytorch/issues/159168


cc @EikanWang @jgong5 @wenzhe-nrv @sanchitintel

malfet committed:
Update

[ghstack-poisoned]

malfet committed:
Update

[ghstack-poisoned]

malfet committed:
Update

[ghstack-poisoned]

malfet committed:
Update

[ghstack-poisoned]

malfet committed:
Update

[ghstack-poisoned]

ezyang commented on aten/src/ATen/core/ivalue.cpp:
```diff
@@ -97,6 +97,9 @@ c10::TypePtr IValue::TagType<c10::Type>::get(const IValue& v) {
       case Tag::Int:
         return IntType::get();
+      case Tag::UInt:
+        TORCH_CHECK(false, "you should not be called");
```
Better message here? is it an INTERNAL_ASSERT?

ezyang reviewed (approved):
Thanks a lot!

albanD reviewed (commented):
I am bit confused on when each tag is used. But I think the overall logic is correct.
Would you be able to add a comment in ivalue.h with details on when unsigned vs signed int are used?

malfet commented on aten/src/ATen/core/ivalue.cpp:
```diff
@@ -377,6 +382,7 @@ size_t IValue::hash(const IValue& v) {
     case Tag::Storage:
       return c10::get_hash(v.payload.u.as_int);
     case Tag::Int:
+    case Tag::UInt:
```
Moved it below the warning

malfet commented on aten/src/ATen/core/ivalue.h:
```diff
@@ -653,6 +655,25 @@ struct TORCH_API IValue final {
   }
 
+  // Unsigned
+  IValue(uint64_t u) : tag( u <= std::numeric_limits<int64_t>::max() ? Tag::Int : Tag::UInt) {
```
`std::numeric_limits<int64_t>::max()` should always be safely castable to `uint64_t`, so yes, feels well defined to me

malfet commented on aten/src/ATen/core/ivalue.h:
```diff
@@ -653,6 +655,25 @@ struct TORCH_API IValue final {
+
+
+  bool isUnsigned() const {
+    return Tag::UInt == tag || (Tag::Int == tag && payload.u.as_int >= 0);;
```
They aren't, same way as as they are for `c10::Scalar`, but let me add a comment there...

malfet committed:
Update on "Add unsigned support to `IValue`"


- Moved repeated logic of saving int64/uint64 into a polymorphic container into `THPUtils_unpackInteger`
- Added `TestPythonDispatch.test_dispatch_uint64` regression test 

Fixes https://github.com/pytorch/pytorch/issues/159168


cc EikanWang jgong5 wenzhe-nrv sanchitintel

[ghstack-poisoned]

malfet commented:
@pytorchbot merge
